beacon_node/beacon_chain/src/block_times_cache.rs

```diff
@@ -0,0 +1,170 @@
+//! was set as head late.
+//! This allows us to distingush between the following scenarios:
+//! - The block was observed late.
+//! - We were too slow to import it.
+//! - We were too slow to set it as head.
+
+use eth2::types::{Hash256, Slot};
+use std::collections::HashMap;
+use std::time::Duration;
+
+type BlockRoot = Hash256;
+
+#[derive(Clone, Default)]
+pub struct Timestamps {
+    pub observed: Option<Duration>,
+    pub imported: Option<Duration>,
+    pub set_as_head: Option<Duration>,
+}
+
+// Helps arrange delay data so it is more relevant to metrics.
+#[derive(Default)]
+pub struct BlockDelays {
+    pub observed: Option<Duration>,
+    pub imported: Option<Duration>,
+    pub set_as_head: Option<Duration>,
+}
+
+impl BlockDelays {
+    fn new(times: Timestamps, slot_start_time: Duration) -> BlockDelays {
+        let observed = times
+            .observed
+            .and_then(|observed_time| observed_time.checked_sub(slot_start_time));
+        let imported = times
+            .imported
+            .and_then(|imported_time| imported_time.checked_sub(times.observed?));
+        let set_as_head = times
+            .set_as_head
+            .and_then(|set_as_head_time| set_as_head_time.checked_sub(times.imported?));
+        BlockDelays {
+            observed,
+            imported,
+            set_as_head,
+        }
+    }
+}
+
+// If the block was received via gossip, we can record the client type of the peer which sent us
+// the block.
+#[derive(Clone, Default)]
+pub struct BlockPeerInfo {
+    pub id: Option<String>,
+    pub client: Option<String>,
+}
+
+pub struct BlockTimesCacheValue {
+    pub slot: Slot,
+    pub timestamps: Timestamps,
+    pub peer_info: BlockPeerInfo,
+}
+
+#[derive(Default)]
+pub struct BlockTimesCache {
+    pub cache: HashMap<BlockRoot, BlockTimesCacheValue>,
+}
+
+/// Helper methods to read from and write to the cache.
+impl BlockTimesCache {
+    pub fn set_time_observed(
+        &mut self,
+        block_root: BlockRoot,
+        slot: Slot,
+        timestamp: Duration,
+        peer_id: Option<String>,
+        peer_client: Option<String>,
+    ) {
+        if let Some(mut block_times) = self.cache.get_mut(&block_root) {
+            block_times.timestamps.observed = Some(timestamp);
+            block_times.peer_info = BlockPeerInfo {
+                id: peer_id,
+                client: peer_client,
+            };
+        } else {
+            let timestamps = Timestamps {
+                observed: Some(timestamp),
+                ..Default::default()
+            };
+            let peer_info = BlockPeerInfo {
+                id: peer_id,
+                client: peer_client,
+            };
+            self.cache.insert(
+                block_root,
+                BlockTimesCacheValue {
+                    slot,
+                    timestamps,
+                    peer_info,
+                },
+            );
+        }
```

Comment:
This pattern is well-suited to `entry(..).or_insert_with(..)`:

```rust
        let block_times = self
            .cache
            .entry(block_root)
            .or_insert_with(|| BlockTimesCacheValue::new(slot));
        block_times.timestamps.observed = Some(timestamp);
        block_times.peer_info = BlockPeerInfo {
            id: peer_id,
            client: peer_client,
        };
```

Where `new` is defined as:

```rust
impl BlockTimesCacheValue {
    fn new(slot: Slot) -> Self {
        BlockTimesCacheValue {
            slot,
            timestamps: Default::default(),
            peer_info: Default::default(),
        }
    }
}
```

Reply:
This is super cool! TIL